samples/apps/copilot-chat-app/webapi/CopilotChat/Skills/ChatSkills/ExternalInformationSkill.cs

```diff
@@ -0,0 +1,324 @@
+        CopilotChatPlanner planner,
+        PlannerOptions plannerOptions)
+    {
+        this._promptOptions = promptOptions.Value;
+        this._plannerOptions = plannerOptions;
+        this._planner = planner;
+    }
+
+    /// <summary>
+    /// Extract relevant additional knowledge using a planner.
+    /// </summary>
+    [SKFunction("Acquire external information")]
+    [SKFunctionName("AcquireExternalInformation")]
+    [SKFunctionInput(Description = "The intent to whether external information is needed")]
+    [SKFunctionContextParameter(Name = "tokenLimit", Description = "Maximum number of tokens")]
+    [SKFunctionContextParameter(Name = "proposedPlan", Description = "Previously proposed plan that is approved")]
+    public async Task<string> AcquireExternalInformationAsync(string userIntent, SKContext context)
+    {
+        if (!this._plannerOptions.Enabled)
+        {
+            return string.Empty;
+        }
+
+        // Skills run in the planner may modify the SKContext. Clone the context to avoid
+        // modifying the original context variables.
+        SKContext plannerContext = Utilities.CopyContextWithVariablesClone(context);
+
+        // Use the user intent message as the input to the plan.
+        plannerContext.Variables.Update(userIntent);
+
+        // Check if plan exists in ask's context variables.
+        // If plan was returned at this point, that means it was approved and should be run
+        var planApproved = context.Variables.Get("proposedPlan", out var planJson);
+
+        if (planApproved && !string.IsNullOrWhiteSpace(planJson))
+        {
+            // Reload the plan with the planner's kernel so
+            // it has full context to be executed
+            var newPlanContext = new SKContext(
+                context.Variables,
+                this._planner.Kernel.Memory,
+                this._planner.Kernel.Skills,
+                this._planner.Kernel.Log
+            );
+            var plan = Plan.FromJson(planJson, newPlanContext);
+
+            // Invoke plan
+            SKContext planContext = await plan.InvokeAsync(plannerContext);
+            int tokenLimit = int.Parse(context["tokenLimit"], new NumberFormatInfo());
+
+            // The result of the plan may be from an OpenAPI skill. Attempt to extract JSON from the response.
+            bool extractJsonFromOpenApi =
+                this.TryExtractJsonFromOpenApiPlanResult(plannerContext, planContext.Variables.Input, out string planResult);
+            if (extractJsonFromOpenApi)
+            {
+                planResult = this.OptimizeOpenApiSkillJson(planResult, tokenLimit, plan);
+            }
+            else
+            {
+                // If not, use result of the plan execution result directly.
+                planResult = planContext.Variables.Input;
+            }
+
+            return $"[START RELATED INFORMATION]\n{planResult.Trim()}\n[END RELATED INFORMATION]\n";
```

Comment:
I don't think it makes a huge difference but may want to account for static strings in the token limit here too

Reply:
I can make the changes now. But eventually we should put these in a config file (Someone open an issue suggesting they would like support for other languages).